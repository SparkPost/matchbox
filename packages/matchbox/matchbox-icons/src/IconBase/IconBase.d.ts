import React from 'react';
interface IconBaseProps extends React.SVGProps<SVGSVGElement> {
    children?: React.ReactNode;
    width?: number | string;
    height?: number | string;
    size?: number | string;
    style?: React.CSSProperties;
    label?: string;
    viewBox?: string;
    text?: string;
    textFill?: string;
    textProps?: React.SVGProps<SVGTextElement>;
}
declare const IconBase: React.FC<IconBaseProps>;
export declare function createSvgIcon(path: JSX.Element, displayName: string): {
    (props: IconBaseProps): JSX.Element;
    displayName: string;
};
export declare function createExtendedSvgIcon({ displayName, path, textContainer, viewBox, }: {
    displayName: string;
    path: JSX.Element;
    textContainer?: object;
    viewBox?: string;
}): {
    ({ text, textFill, textProps, ...props }: IconBaseProps): JSX.Element;
    displayName: string;
};
export default IconBase;
