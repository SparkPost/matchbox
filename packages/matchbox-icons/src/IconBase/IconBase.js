import React, { Component } from 'react';
import PropTypes from 'prop-types';

class IconBase extends Component {
  static displayName = 'IconBase';

  static defaultProps = {
    size: 18
  };

  static propTypes = {
    size: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]),
    width: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]),
    height: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]),
    label: PropTypes.string,
    viewBox: PropTypes.string,
    /** FileType icon only */
    text: PropTypes.string,
    /** FileType icon only */
    textFill: PropTypes.string,
    /** FileType icon only */
    textProps: PropTypes.object
  }

  render() {
    const {
      children,
      width,
      height,
      size,
      style = {},
      viewBox = '0 0 24 24',
      label,
      ...rest
    } = this.props;

    const styleProp = {
      verticalAlign: 'middle',
      ...style
    };

    return (
      <svg
        fill='currentColor'
        preserveAspectRatio='xMidYMid meet'
        viewBox={viewBox}
        width={width || size}
        height={height || size}
        aria-label={label}
        role={label && 'img'}
        aria-hidden={label ? 'false' : 'true'}
        {...rest}
        style={styleProp}>
        {children}
      </svg>
    );
  }
}

export function createSvgIcon(path, displayName) {
  const Icon = (props) => (
    <IconBase {...props}>{path}</IconBase>
  );

  Icon.displayName = displayName;

  return Icon;
}

export function createExtendedSvgIcon({
  displayName,
  path,
  text,
  textContainer,
  viewBox
}) {
  const Icon = ({
    text,
    textFill = 'white',
    textProps,
    ...props
  }) => (
    <IconBase viewBox={viewBox} {...props}>
      {path}
      <text fill={textFill} {...textContainer} {...textProps}>
        {text}
      </text>
    </IconBase>
  );

  Icon.displayName = displayName;

  return Icon;
}

export default IconBase;
