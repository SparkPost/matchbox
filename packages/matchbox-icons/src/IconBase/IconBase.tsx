import React from 'react';

export type IconBaseProps = React.ComponentPropsWithoutRef<'svg'> & {
  children?: React.ReactNode;
  width?: number | string;
  height?: number | string;
  size?: number | string;
  label?: string;
  viewBox?: string;
  text?: string;
  textFill?: string;
  textProps?: React.ComponentPropsWithoutRef<'text'>;
};

const IconBase = (props: IconBaseProps): JSX.Element => {
  const {
    children,
    width,
    height,
    size = 18,
    style,
    viewBox = '0 0 24 24',
    label,
    ...rest
  } = props;

  const styleProp = {
    verticalAlign: 'middle',
    ...style,
  };

  return (
    <svg
      fill="currentColor"
      preserveAspectRatio="xMidYMid meet"
      viewBox={viewBox}
      width={width || size}
      height={height || size}
      aria-label={label}
      role={label && 'img'}
      aria-hidden={label ? 'false' : 'true'}
      {...rest}
      style={styleProp}
    >
      {children}
    </svg>
  );
};

IconBase.displayName = 'IconBase';

export function createSvgIcon(path: JSX.Element, displayName: string): React.ElementType {
  const Icon = (props: IconBaseProps) => <IconBase {...props}>{path}</IconBase>;

  Icon.displayName = displayName;

  return Icon;
}

export function createExtendedSvgIcon({
  displayName,
  path,
  textContainer,
  viewBox,
}: {
  displayName: string;
  path: JSX.Element;
  textContainer?: {
    [k: string]: unknown;
  };
  viewBox?: string;
}): React.ElementType {
  const Icon = ({ text, textFill = 'white', textProps, ...props }: IconBaseProps) => (
    <IconBase viewBox={viewBox} {...props}>
      <g>
        {path}
        <text fill={textFill} {...textContainer} {...textProps}>
          {text}
        </text>
      </g>
    </IconBase>
  );

  Icon.displayName = displayName;

  return Icon;
}

export default IconBase;
