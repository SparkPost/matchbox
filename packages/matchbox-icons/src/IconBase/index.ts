export * from './IconBase';
export { default as IconBase } from './IconBase';
