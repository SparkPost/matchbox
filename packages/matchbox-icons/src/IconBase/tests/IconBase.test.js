import React from 'react';
import IconBase, { createSvgIcon, createExtendedSvgIcon } from '../IconBase';
import { shallow } from 'enzyme';

describe('IconBase', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<IconBase><g><path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z" /></g></IconBase>);
  });

  it('renders children correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders size correctly', () => {
    wrapper.setProps({ size: 10 });
    expect(wrapper.instance().props.size).toEqual(10);

    wrapper.setProps({ size: null, width: 16, height: 15 });
    expect(wrapper.instance().props.width).toEqual(16);
    expect(wrapper.instance().props.height).toEqual(15);
  });

  it('renders an accessible label when the label prop has a value', () => {
    wrapper.setProps({ label: 'Hello, world' });

    expect(wrapper.instance().props.label).toEqual('Hello, world');
    expect(wrapper).toMatchSnapshot();
  });
});

describe('createSvgIcon', () => {
  it('creates an Icon correctly', () => {
    const Icon = createSvgIcon(<path d="a path" />, 'TestIcon');
    expect(shallow(<Icon />)).toMatchSnapshot();
  });
});

describe('createExtendedSvgIcon', () => {
  it('creates an extended Icon correctly', () => {
    const Icon = createExtendedSvgIcon({
      path: <path d="a path"/>,
      displayName: 'TextIcon',
      viewBox: '0 0 100 100',
      textContainer: {
        x: '22',
        y: '62',
        fontSize: '24'
      }
    });
    expect(shallow(
      <Icon text='test text' textFill='white' textProps={{ stroke: 'black' }} />
    )).toMatchSnapshot();
  });
});
