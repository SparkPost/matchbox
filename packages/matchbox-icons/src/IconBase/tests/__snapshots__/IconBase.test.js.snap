// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconBase renders an accessible label when the label prop has a value 1`] = `
<svg
  aria-hidden="false"
  aria-label="Hello, world"
  fill="currentColor"
  height={18}
  preserveAspectRatio="xMidYMid meet"
  role="img"
  style={
    Object {
      "verticalAlign": "middle",
    }
  }
  viewBox="0 0 24 24"
  width={18}
>
  <g>
    <path
      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
    />
  </g>
</svg>
`;

exports[`IconBase renders children correctly 1`] = `
<svg
  aria-hidden="true"
  fill="currentColor"
  height={18}
  preserveAspectRatio="xMidYMid meet"
  style={
    Object {
      "verticalAlign": "middle",
    }
  }
  viewBox="0 0 24 24"
  width={18}
>
  <g>
    <path
      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
    />
  </g>
</svg>
`;

exports[`createExtendedSvgIcon creates an extended Icon correctly 1`] = `
<IconBase
  size={18}
  viewBox="0 0 100 100"
>
  <path
    d="a path"
  />
  <text
    fill="white"
    fontSize="24"
    stroke="black"
    x="22"
    y="62"
  >
    test text
  </text>
</IconBase>
`;

exports[`createSvgIcon creates an Icon correctly 1`] = `
<IconBase
  size={18}
>
  <path
    d="a path"
  />
</IconBase>
`;
